I can't reproduce those fails:
```
[ERROR] Failures: 
[ERROR]   CDIPlainInjectionTest>Arquillian.arquillianBeforeClass:109 » Deployment Except...
[ERROR]   CDIPropertyNameMatchingTest>Arquillian.arquillianBeforeClass:109 » Deployment ...
[ERROR]   ConfigProviderTest>Arquillian.run:164->testEnvironmentConfigSource:97 » StringIndexOutOfBounds
```
they pass for me. It seems there's something with #getListSize() on certain environments, can you check that those fail/pass for you?

--- a/.gitignore
+++ b/.gitignore
@@ -157,3 +157,4 @@ $RECYCLE.BIN/
 *.lnk
 
 
+/test-output/

--- a/pom.xml
+++ b/pom.xml
@@ -22,7 +22,7 @@
         <maven.compiler.target>1.8</maven.compiler.target>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 
-        <kumuluzee.version>3.0.0</kumuluzee.version>
+        <kumuluzee.version>3.5.0</kumuluzee.version>
 
         <microprofile-config.version>1.3</microprofile-config.version>
         <hamcrest.version>1.3</hamcrest.version>

--- a/src/main/java/com/kumuluz/ee/config/microprofile/ConfigImpl.java
+++ b/src/main/java/com/kumuluz/ee/config/microprofile/ConfigImpl.java
@@ -20,20 +20,25 @@
 */
 package com.kumuluz.ee.config.microprofile;
 
-import com.kumuluz.ee.config.microprofile.converters.ImplicitConverter;
-import com.kumuluz.ee.config.microprofile.utils.AlternativeTypesUtil;
-import org.eclipse.microprofile.config.Config;
-import org.eclipse.microprofile.config.spi.ConfigSource;
-import org.eclipse.microprofile.config.spi.Converter;
-
 import java.io.Serializable;
 import java.lang.reflect.Array;
 import java.lang.reflect.Type;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
 import java.util.logging.Logger;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.eclipse.microprofile.config.spi.Converter;
+
+import com.kumuluz.ee.config.microprofile.converters.ImplicitConverter;
+import com.kumuluz.ee.config.microprofile.utils.AlternativeTypesUtil;
 
 /**
  * Microprofile Config implementation that exposes KumuluzEE configuration framework.
@@ -164,12 +169,6 @@ public class ConfigImpl implements Config, Serializable {
 	 * @return stream of property names, never null
 	 */
 	private Stream<String> asStream(ConfigSource arg0) {
-		try {
-			return arg0.getPropertyNames().stream();
-		}
-		catch (Exception e) {
-			logger.warning("Config source " + arg0.getName() + " reported exception: " + e.getMessage());
-			return Collections.EMPTY_SET.stream();
-		}
+		return arg0.getPropertyNames().stream();
 	}
 }

--- a/src/main/java/com/kumuluz/ee/config/microprofile/ConfigImpl.java
+++ b/src/main/java/com/kumuluz/ee/config/microprofile/ConfigImpl.java
@@ -31,7 +31,6 @@ import java.util.Optional;
 import java.util.logging.Logger;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
-import java.util.stream.Stream;
 
 import org.eclipse.microprofile.config.Config;
 import org.eclipse.microprofile.config.spi.ConfigSource;
@@ -155,20 +154,11 @@ public class ConfigImpl implements Config, Serializable {
 
 	@Override
 	public Iterable<String> getPropertyNames() {
-		return this.configSources.stream().flatMap(this::asStream).collect(Collectors.toSet());
+		return this.configSources.stream().flatMap(e -> e.getPropertyNames().stream()).collect(Collectors.toSet());
 	}
 
 	@Override
 	public Iterable<ConfigSource> getConfigSources() {
 		return this.configSources;
-	}
-
-	/**
-	 * Streams this config source as property names
-	 * @param arg0 config source to stream
-	 * @return stream of property names, never null
-	 */
-	private Stream<String> asStream(ConfigSource arg0) {
-		return arg0.getPropertyNames().stream();
 	}
 }

--- a/src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java
+++ b/src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java
@@ -24,7 +24,6 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-import java.util.stream.Collectors;
 
 import org.eclipse.microprofile.config.spi.ConfigSource;
 
@@ -61,12 +60,15 @@ public class ConfigSourceAdapter implements ConfigSource {
     @Override
     public String getValue(String s) {
     	String val = configurationSource.get(s).orElse(null);
-        Optional<Integer> listSize = this.configurationSource.getListSize(s);
-        
-        //this is a list or an array
-    	if (listSize.isPresent()) {
-    		//we ignore the returned value and build the array
-            return buildArray(s, listSize.get());
+    	
+    	if (val != null) {
+	        Optional<Integer> listSize = this.configurationSource.getListSize(s);
+	        
+	        //this is a list or an array
+	    	if (listSize.isPresent()) {
+	    		//we ignore the returned value and build the array
+	            return buildArray(s, listSize.get());
+	    	}
     	}
     	
         return val;
@@ -112,16 +114,9 @@ public class ConfigSourceAdapter implements ConfigSource {
 			String prefix = String.format("%s[%d]", propertyName, i);
 	        Optional<List<String>> objectKeys = this.configurationSource.getMapKeys(prefix);
 
-	        //array item is an object, so we just toString() it
+	        //array item is an object, so we just omit it
 	        if (objectKeys.isPresent()) {
-	            Map<String, String> map = new HashMap<>();
-	        	buildPropertiesMap(map, prefix);
-	        	
-	    		Map<String, String> object =
-	    				map.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().substring(prefix.length() + 1),
-	    						e -> e.getValue()));
-	        	
-	        	sb.append(object.toString().replaceAll(",", "\\\\,"));
+	        	sb.append("");
 	        }
 	        else {
 				Optional<String> item = this.configurationSource.get(String.format("%s[%d]", propertyName, i));

--- a/src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java
+++ b/src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java
@@ -24,7 +24,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 import org.eclipse.microprofile.config.spi.ConfigSource;
 
@@ -34,11 +34,10 @@ import com.kumuluz.ee.configuration.ConfigurationSource;
  * Adapts KumuluzEE configuration framework {@link ConfigurationSource} to MicroProfile Config {@link ConfigSource}.
  *
  * @author Urban Malc
+ * @author Yog Sothoth
  * @since 1.3
  */
 public class ConfigSourceAdapter implements ConfigSource {
-    private static String ARRAY_SEPARATOR_REGEX = "(?<!\\\\)" + Pattern.quote(",") + "\\s";    
-
     private ConfigurationSource configurationSource;
 
     public ConfigSourceAdapter(ConfigurationSource configurationSource) {
@@ -62,10 +61,12 @@ public class ConfigSourceAdapter implements ConfigSource {
     @Override
     public String getValue(String s) {
     	String val = configurationSource.get(s).orElse(null);
-    	
-    	if (configurationSource.getListSize(s).orElse(0) > 0) {
-    		//this is a list or an array and leading/trailing "[", "]" as well as \\s after "," must be removed
-    		return val.subSequence(1, val.length() - 1).toString().replaceAll(ARRAY_SEPARATOR_REGEX, ",");
+        Optional<Integer> listSize = this.configurationSource.getListSize(s);
+        
+        //this is a list or an array
+    	if (listSize.isPresent()) {
+    		//we ignore the returned value and build the array
+            return buildArray(s, listSize.get());
     	}
     	
         return val;
@@ -103,4 +104,35 @@ public class ConfigSourceAdapter implements ConfigSource {
             }
         }
     }
+	
+	private String buildArray(String propertyName, int size) {
+		StringBuilder sb = new StringBuilder();
+
+		for (int i = 0; i < size; i++) {
+			String prefix = String.format("%s[%d]", propertyName, i);
+	        Optional<List<String>> objectKeys = this.configurationSource.getMapKeys(prefix);
+
+	        //array item is an object, so we just toString() it
+	        if (objectKeys.isPresent()) {
+	            Map<String, String> map = new HashMap<>();
+	        	buildPropertiesMap(map, prefix);
+	        	
+	    		Map<String, String> object =
+	    				map.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().substring(prefix.length() + 1),
+	    						e -> e.getValue()));
+	        	
+	        	sb.append(object.toString().replaceAll(",", "\\\\,"));
+	        }
+	        else {
+				Optional<String> item = this.configurationSource.get(String.format("%s[%d]", propertyName, i));
+				item.ifPresent(sb::append);
+	        }
+			
+			if (i < size - 1) {
+				sb.append(',');
+			}
+		}
+		
+		return sb.toString();
+	}	
 }

--- a/src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java
+++ b/src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java
@@ -76,7 +76,7 @@ public class ArrayInjectionTest extends Arquillian {
 	private String arrayItemParam;
 	
 	private static final String[] TEST_TARGET =
-			new String[]{"one ", " two", " [three, four] ", "{five.six=seven, five.eight=nine}"};
+			new String[]{"one ", " two", " [three, four] "};
 	
     @Test
     public void testNotParsingDelimeters() {

--- a/src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java
+++ b/src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java
@@ -1,0 +1,65 @@
+package com.kumuluz.ee.config.microprofile.tests;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+@Test
+public class ArrayInjectionTest extends Arquillian {
+    @Deployment
+    public static JavaArchive deploy() {
+        JavaArchive testJar = ShrinkWrap
+                .create(JavaArchive.class, "arrayInjectionTest.jar")
+                .addClasses(ArrayInjectionTest.class)
+                .addAsResource("config.yaml")
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+                .as(JavaArchive.class);
+            return testJar;
+    }
+
+	@Inject
+	@ConfigProperty(name="parameter.arrayParameter")
+	private List<String> listParam;
+
+	@Inject
+	@ConfigProperty(name="parameter.arrayParameter")
+	private String[] arrayParam;
+
+	@Inject
+	@ConfigProperty(name="parameter.arrayParameter")
+	private Optional<String[]> arrayParamOpt;
+
+	@Inject
+	@ConfigProperty(name="parameter.stringParameter")
+	private Optional<String> stringParam;
+
+	private static final String[] TEST_TARGET =
+			new String[]{"one ", " two", " [three, four] "};
+	
+    @Test
+    public void testNotParsingDelimeters() {
+		Assert.assertEquals("[here be,  dragons]", stringParam.get());
+    }
+
+    @Test
+    public void testArrayInjection() {
+		Assert.assertEquals(TEST_TARGET, arrayParamOpt.get());
+		Assert.assertEquals(TEST_TARGET, arrayParam);
+    }
+    
+    @Test
+    public void testListInjection() {
+		Assert.assertEquals(Arrays.asList(TEST_TARGET), listParam);
+    }
+}

--- a/src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java
+++ b/src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java
@@ -1,3 +1,23 @@
+/*
+ *  Copyright (c) 2014-2017 Kumuluz and/or its affiliates
+ *  and other contributors as indicated by the @author tags and
+ *  the contributor list.
+ *
+ *  Licensed under the MIT License (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  https://opensource.org/licenses/MIT
+ *
+ *  The software is provided "AS IS", WITHOUT WARRANTY OF ANY KIND, express or
+ *  implied, including but not limited to the warranties of merchantability,
+ *  fitness for a particular purpose and noninfringement. in no event shall the
+ *  authors or copyright holders be liable for any claim, damages or other
+ *  liability, whether in an action of contract, tort or otherwise, arising from,
+ *  out of or in connection with the software or the use or other dealings in the
+ *  software. See the License for the specific language governing permissions and
+ *  limitations under the License.
+*/
 package com.kumuluz.ee.config.microprofile.tests;
 
 import java.util.Arrays;
@@ -15,6 +35,13 @@ import org.jboss.shrinkwrap.api.spec.JavaArchive;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
+/**
+ * Tests that check MP Config array/list injection from KumuluzEE's config.yaml
+ * 
+ * @author Yog Sothoth
+ * @since 1.4
+ *
+ */
 @Test
 public class ArrayInjectionTest extends Arquillian {
     @Deployment
@@ -44,12 +71,21 @@ public class ArrayInjectionTest extends Arquillian {
 	@ConfigProperty(name="parameter.stringParameter")
 	private Optional<String> stringParam;
 
+	@Inject
+	@ConfigProperty(name="parameter.arrayParameter[0]")
+	private String arrayItemParam;
+	
 	private static final String[] TEST_TARGET =
-			new String[]{"one ", " two", " [three, four] "};
+			new String[]{"one ", " two", " [three, four] ", "{five.six=seven, five.eight=nine}"};
 	
     @Test
     public void testNotParsingDelimeters() {
 		Assert.assertEquals("[here be,  dragons]", stringParam.get());
+    }
+
+    @Test
+    public void testArrayItemInjection() {
+    	Assert.assertEquals(TEST_TARGET[0], arrayItemParam);
     }
 
     @Test

--- a/src/test/resources/config.yaml
+++ b/src/test/resources/config.yaml
@@ -4,3 +4,6 @@ parameter:
     - "one "
     - " two"
     - " [three\\, four] "
+    - five:
+        six: seven
+        eight: nine

--- a/src/test/resources/config.yaml
+++ b/src/test/resources/config.yaml
@@ -1,0 +1,6 @@
+parameter:
+  stringParameter: "[here be,  dragons]"
+  arrayParameter:
+    - "one "
+    - " two"
+    - " [three\\, four] "

